sycl/source/detail/kernel_impl.hpp

```diff
@@ -247,7 +247,7 @@ class kernel_impl {
   const DeviceImageImplPtr MDeviceImageImpl;
   const KernelBundleImplPtr MKernelBundleImpl;
   bool MIsInterop = false;
-  std::mutex MNoncacheableEnqueueMutex;
+  mutable std::mutex MNoncacheableEnqueueMutex;
```

Comment:
I'm not sure what's our approach for `const`, we probably didn't care before because shared pointers could always give us non-const ptr even if they themselves were const.

@sergey-semenov , any preferences? I think the pointer we get below is non-const, so we can avoid this bit by simply changing the signature to accept non-const.

Reply:
My preference is that if the object itself is not modified, then a `const` pointer should be used and use `mutable std::mutex` since it might be required for read-only concurrent access synchronization.

Reply:
My issue is that we (most likely) don't do that consistently. I'd rather have "wrong" behavior everywhere than the "right" one somewhere only. But this isn't big deal for me.

Reply:
> we probably didn't care before because shared pointers could always give us non-const ptr even if they themselves were const.

That's probably true in some (most?) places. I don't have a strong opinion on this, but without knowing exactly how widespread the problem is, I'd rather not contribute to it and start fixing the pieces that we hit, so I'm leaning to sticking with `mutable` here.